### ISSUE TITLE
Randomly blow up memory usage and crash sometimes when under heavy usage

### ISSUE THREAD

**Issue by xzn** [piroor/treestyletab#2340]
On Firefox 68 on Windows 10 x64

I don't really have a way of reproducing (not sure if it's even an issue with this addon)

Sometimes under heavy usage TST sidebar would complete freeze, gigabytes of memory would get used and eventually Firefox ask me to kill some "unresponsive tab". The sidebar falls into a zombie state afterward and won't function again until I disabled and reenabled the addon.

Console log shows a crash in `common.js`. However Firefox likes to randomly blank out the developer tools window and then all the log are lost so I don't have anything to submit from there.

If anyone know how to make the developer tool window not randomly crash (as in it turns blank and as far as I can tell can only be brought back by closing it and reopening it from `about:debugging`, at which point all logs are gone, even though `persist logs` is turn on.) Please let me know.

In short:

TST randomly crashes and blow up memory usage. I can't submit logs because `persist logs` doesn't work..

**Comment by xzn** [comment]
Apparently logs for an extensions is lost if you disable/reenable it even if `persist logs` is checked.

In any case I get the following:
```
Error: Script terminated by timeout at:
updateAncestors@moz-extension://6139028a-128d-4687-83a5-cbc641b83c02/common/Tab.js:477:12
get ancestors@moz-extension://6139028a-128d-4687-83a5-cbc641b83c02/common/Tab.js:469:19
updateAncestors@moz-extension://6139028a-128d-4687-83a5-cbc641b83c02/background/background.js:395:1
reserveToUpdateAncestors/tab.$TST.reservedUpdateAncestors<@moz-extension://6139028a-128d-4687-83a5-cbc641b83c02/background/background.js:383:7
```

This is from the version of TST on AMO. The line seems to be from [here](https://github.com/piroor/treestyletab/blob/44d9a6458a1536e773fadb59e6cdc6e60128f503/webextensions/common/Tab.js#L477)

Not sure if the problem really is here though, the loop itself seems harmless.

**Comment by piroor** [comment]
This type stack trace possibly indicate a non-bottleneck code, because it just indicates that which line is executed when the process is stopped. Instead [a performance profile](https://github.com/piroor/treestyletab/wiki/How-to-inspect-tree-of-tabs#how-to-collect-a-performance-profile) may reveal what the bottleneck is.

**Comment by xzn** [comment]
I'll give it a try. As mentioned above it only happens sporadically and I have no way of reproducing it properly.

Does the performance profiling takes up a lot of space? I've tried GC/CC log before and it uses like 10GB of space in a few minutes.

In the performance profiling settings, are the default options good enough?
![image](https://user-images.githubusercontent.com/1617680/62034333-399fef80-b1c4-11e9-9073-1e178930217e.png)

**Comment by xzn** [comment]
After some logging, there appears to be cyclical references of `parent` and `child`. So `get ancestors()` and `get descendants()` go into infinite loop. Don't know why it happens though.

**Comment by piroor** [comment]
Thanks. I thought that TST has some guards for such a cyclic reference, but there looks to be some unhandled cases. Hmm...

Do you use drag and drop to modify tree structure frequently? Or, is the problem happen on opening/closing of tabs?

**Comment by xzn** [comment]
So far it only happened on opening/closing of tabs. I've no way of reproducing it consistently however.

**Comment by piroor** [comment]
I've introduce new guard for cyclic references. I hope it covers your case.

**Comment by xzn** [comment]
Thanks for the update! I'll reopen this if I run into this issue again in the future.

**Comment by xzn** [comment]
The crash happened again:

```
TSTAPI.initAsBackend: failed to send "ready" message to "multipletab@piro.sakura.ne.jp": Error: "Could not establish connection. Receiving end does not exist."
    id moz-extension://b3a87a13-c533-4311-87d1-22bf7a0b3728/common/tst-api.js:517
tst-api.js:532:15
FATAL ERROR: Failed to get unique id for a tab 51:  Error: "Invalid tab ID: 51" Tab.js:207:15
FATAL ERROR: Failed to get unique id for a tab 64:  Error: "Invalid tab ID: 64" Tab.js:207:15
TreeStructure.reserveToAttachTabFromRestoredInfo: Fatal error on processing task 38, Error: Invalid tab ID: 64 tree-structure.js:155:17
FATAL ERROR: Failed to get unique id for a tab 76:  Error: "Invalid tab ID: 76" Tab.js:207:15
TreeStructure.reserveToAttachTabFromRestoredInfo: Fatal error on processing task 37, Error: Invalid tab ID: 76 tree-structure.js:155:17
FATAL ERROR: Failed to get unique id for a tab 230:  Error: "Invalid tab ID: 230" Tab.js:207:15
Error: "Tab.waitUntilTracked for 303 is timed out (in bg)false"
    timeout moz-extension://b3a87a13-c533-4311-87d1-22bf7a0b3728/common/Tab.js:1157
api-tabs-listener.js:201:13
Error: "Tab.waitUntilTracked for 303 is timed out (in 3)false"
    timeout moz-extension://b3a87a13-c533-4311-87d1-22bf7a0b3728/common/Tab.js:1157
EventListenerManager.js:65:23
Error: "Tab.waitUntilTracked for 312 is timed out (in bg)false"
    timeout moz-extension://b3a87a13-c533-4311-87d1-22bf7a0b3728/common/Tab.js:1157
api-tabs-listener.js:201:13
Error: "Tab.waitUntilTracked for 312 is timed out (in 3)false"
    timeout moz-extension://b3a87a13-c533-4311-87d1-22bf7a0b3728/common/Tab.js:1157
EventListenerManager.js:65:23
Error: "Tab.waitUntilTracked for 314 is timed out (in bg)false"
    timeout moz-extension://b3a87a13-c533-4311-87d1-22bf7a0b3728/common/Tab.js:1157
api-tabs-listener.js:201:13
Error: "Tab.waitUntilTracked for 314 is timed out (in 3)false"
    timeout moz-extension://b3a87a13-c533-4311-87d1-22bf7a0b3728/common/Tab.js:1157
EventListenerManager.js:65:23
```

Though it crashed more quickly this time though so I guess the guard is working.

**Comment by piroor** [comment]
It looks that multiple tabs are closed while TST's initialization process. Is there any possible reason on the startup process of your Firefox, for example another add-on? Or, do you use the private browsing mode and you granted TST to access tabs in private windows?

**Comment by xzn** [comment]
Usually this happens after opening a bunch of tabs, closing a bunch of tabs, and reopening a bunch of tabs in quick successions. Happens in private browsing mode as well but happens normally just as well.

The only other tab related addon that is used is Tab Session Manager.

**Comment by xzn** [comment]
The crash just happened again.

The guard at d033e67eb7871cfb09ceccacfbc99a55803306c9 appears to not be working. Perhaps the `ancestorsOfSelf` check is not working as it should due to `ensureLivingTab` filter?

**Comment by xzn** [comment]
This should fixes the hang issue:

```diff
--- a/webextensions/background/tree.js
+++ b/webextensions/background/tree.js
@@ -99,7 +99,7 @@ export async function attachTabTo(child, parent, options = {}) {
     log('=> could not attach tab to a parent in different window');
     return false;
   }
-  const ancestors = [parent].concat(child.$TST.ancestors);
+  const ancestors = [parent].concat(parent.$TST.ancestors);
   if (ancestors.includes(child)) {
     log('=> canceled for recursive request');
     return false;
diff --git a/webextensions/common/Tab.js b/webextensions/common/Tab.js
index e7e9cddc..4fd3894f 100644
--- a/webextensions/common/Tab.js
+++ b/webextensions/common/Tab.js
@@ -537,20 +537,20 @@ export default class Tab {
   }

   set children(tabs) {
-    const ancestorsOfSelf = this.ancestors;
-    tabs = tabs.filter(tab => {
-      if (!ancestorsOfSelf.includes(tab))
+    const ancestorsIds = this.ancestors.map(tab => tab.id);
+    const newChildIds = tabs.
+    map(tab => typeof tab == 'number' ? tab : tab && tab.id).
+    filter(id => {
+      if (!ancestorsIds.includes(id))
         return true;
       console.log('FATAL ERROR: Cyclic tree structure has detected and prevented. ', {
-        ancestorsOfSelf,
+        ancestorsIds,
         tabs,
-        tab,
+        id,
         stack: new Error().stack
       });
       return false;
     });
-
-    const newChildIds = tabs.map(tab => typeof tab == 'number' ? tab : tab && tab.id).filter(id => id);
     if (newChildIds.join('|') == this.childIds.join('|'))
       return tabs;
```

Tabs still get reopened in the wrong order though given the steps in #2366

**Comment by piroor** [comment]
@xzn Thanks! I missed that the guard didn't work as expected. And the second change including optimization is also applied in a method based on `Array.prototype.reduce()`. See also: https://github.com/piroor/treestyletab/issues/2278#issuecomment-524961749

**Comment by knobhead** [comment]
Hi, thanks for investigating!

Any ETA in sight for a release  fixing this issue? It is a pretty annoying bug in multi-device ff-sync'ed scenarios as every  tab sent to another device crashes browser (and system, by chance)...

cheers

**Comment by irvinm** [comment]
@knobhead have you can try the nightly build to see if it is resolved for you.  https://www.piro.sakura.ne.jp/xul/xpi/nightly/
